### PR TITLE
Add KPRepublic BM40HSRGB keyboard

### LINKED ISSUE #9463: BM40HSRGB Support[Feature Request]
I would like to add support for the KPRepublic Hotswap BM40RGB PCB.

## Description

The BM40RGB has kbfirmware support. I'd like to port it to qmk firmware. My plan was to use https://noroadsleft.github.io/kbf_qmk_converter/, if needed fix it up so it passes the coding guidlines and add a default keymap.

Would that be acceptable?

### PR DESCRIPTION
**Pull request by tominabox1**
## Description

The current sale page for the BM40 references building firmware from kbfirmware.com and is really confusing so I just figured I'd take care of the port to the master line. I don't own one of these but helped someone set up their board and it appears to work as intended with this code set. 

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Fixes #9463

## Checklist

- [x ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/bm40hsrgb/readme.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+Keyboard Maintainer: [tominabox1](https://github.com/tominabox1)  
+Hardware Supported: bm40hsrgb PCB    
+Hardware Availability: [KPRepublic](https://www.aliexpress.com/item/4001147779116.html)  
+
+Make example for this keyboard (after setting up your build environment):
+
+    make bm40hsrgb:default
+
+See the [build environment setup](https://docs.qmk.fm/#/getting_started_build_tools) and the [make instructions](https://docs.qmk.fm/#/getting_started_make_guide) for more information. Brand new to QMK? Start with our [Complete Newbs Guide](https://docs.qmk.fm/#/newbs).
```
Comment (zvecr) [inline keyboards/bm40hsrgb/readme.md:16]:
```suggestion
Make example for this keyboard (after setting up your build environment):

    make bm40hsrgb:default

Flashing example for this keyboard:

    make bm40hsrgb:default:flash

See the [build environment setup](https://docs.qmk.fm/#/getting_started_build_tools) and the [make instructions](https://docs.qmk.fm/#/getting_started_make_guide) for more information. Brand new to QMK? Start with our [Complete Newbs Guide](https://docs.qmk.fm/#/newbs).
```

File: keyboards/bm40hsrgb/config.h [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+/* key matrix pins */
+#define MATRIX_ROW_PINS { B3, B2, E6, B5 }
+#define MATRIX_COL_PINS { B6, C6, B4, D7, D4, D6, C7, F6, F5, F4, F1, F0 }
+#define UNUSED_PINS
+
+/* COL2ROW or ROW2COL */
+#define DIODE_DIRECTION COL2ROW
+
+/* number of backlight levels */
+
```
Comment (tzarc) [inline keyboards/bm40hsrgb/config.h:26]:
```suggestion
```

File: keyboards/bm40hsrgb/rules.mk [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes           # USB Nkey Rollover
+RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
+BLUETOOTH_ENABLE = no       # Enable Bluetooth
+AUDIO_ENABLE = no           # Audio output
```
Comment (zvecr) [inline keyboards/bm40hsrgb/rules.mk:18]:
```suggestion
# MCU name
MCU = atmega32u4

# Bootloader selection
BOOTLOADER = atmel-dfu

# Build Options
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = no            # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output

LAYOUTS = planck_mit
```

File: keyboards/bm40hsrgb/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    KC_GRV,  KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_BSPC,
+    KC_DEL,  KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,   KC_MINS, KC_EQL,  KC_LBRC, KC_RBRC, KC_BSLS,
+    _______, KC_F7,   KC_F8,   KC_F9,   KC_F10,  KC_F11,  KC_F12,  KC_NUHS, KC_NUBS, KC_PGUP, KC_PGDN, _______,
+    _______, _______, _______, _______, _______, _______, _______, KC_MNXT, KC_VOLD, KC_VOLU, KC_MPLY
+),
+
+[_ADJUST] = LAYOUT_planck_mit(
+    _______, RESET,   DEBUG,   RGB_TOG, RGB_MOD, RGB_HUI, RGB_HUD, RGB_SAI, RGB_SAD,  RGB_VAI, RGB_VAD, KC_DEL ,
+    _______, _______, MU_MOD,  AU_ON,   AU_OFF,  AG_NORM, AG_SWAP, _______,  _______,  _______, _______, _______,
+    _______, MUV_DE,  MUV_IN,  MU_ON,   MU_OFF,  MI_ON,   MI_OFF,  TERM_ON, TERM_OFF, _______, _______, _______,
```
Comment (zvecr) [inline keyboards/bm40hsrgb/keymaps/default/keymap.c:72]:
```suggestion
    _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
    _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______
```

**Comment by tominabox1** [comment]
> Also, we were kinda waiting for a port would include rgb matrix rather than rgb light. (Though maybe at this point "something is better than nothing"?)

I would love to do that but since I don't actually have the board in hand, and I'm not sure how the matrix is set up, it'd be pretty tough to do.

File: keyboards/bm40hsrgb/config.h [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+#define UNUSED_PINS
+
+/* COL2ROW or ROW2COL */
+#define DIODE_DIRECTION COL2ROW
+
+/* number of backlight levels */
+#ifdef BACKLIGHT_PIN
+#define BACKLIGHT_LEVELS 3
+#endif
+
```
Comment (zvecr) [inline keyboards/bm40hsrgb/config.h:43]:
```suggestion
```

**Comment by ncwhale** [comment]
> Also, we were kinda waiting for a port would include rgb matrix rather than rgb light. (Though maybe at this point "something is better than nothing"?)

RGB Matrix is just another DISPLAY port, though it uses digit output driver for differnet point.

Maybe I will get it done next keyboard(need powerful chips for this job.

Atmega32U4 can only process 60fps BW image real time.

**Comment by zvecr** [comment]
@ncwhale I think you misunderstand. Rgb matrix works with ws2812 as well as ISSI based boards. Plenty of boards in the repo already have it enabled, though it does have some timer related shortcomings on AVR chips right now.

File: keyboards/bm40hsrgb/info.json [hunk 6]
```diff
@@ -0,0 +1,61 @@
+{
+    "keyboard_name": "BM40HSRGB",
+    "url": "",
+    "maintainer": "qmk",
+    "width": 12,
+    "height": 4,
+    "layouts": {
+        "LAYOUT_planck_mit": {
+            "key_count": 47,
```
Comment (fauxpark) [inline keyboards/bm40hsrgb/info.json:9]:
```suggestion
```

**Comment by noroadsleft** [comment]
Thanks!

**Comment by tominabox1** [comment]
Awesome work! Thanks for the quick turn.

**Pull request merged**